### DISCUSSION TITLE
Issues and PRs by labels

### DISCUSSION THREAD

**Discussion by alvarosabu** [Tresjs/tres#603] [General]
We are reaching a point where we need a proper prioritisation and categorisation workflow for new features and bugs across the ecosystem.

I got some advice from @patak-dev on how they label issues and PRs https://github.com/vitejs/vite/discussions/15857, I would like to apply something similar:

## Feedback
- `experimental`: Experimental feature

## Priorities
- `pending-triage`: ticket is pending to be prioritized 
- `p1-chore`: Doesn't change end-user code behavior (Ex. new linter, internal playground stuff)

### Bugs

- `p2-edge-case`: Bug, but has a workaround or limited in scope
- `p3-minor-bug`: An edge case that only affects very specific usage
- `p4-important`: Violate documented behavior or significantly improve performance
- `p5-urgent`: Fix build-breaking bugs affecting most users, should be released ASAP

### Features

- `p2-nice-to-have`: Not breaking anything but nice to have
- `p2-to-be-discussed`: Enhancement under consideration
- `p3-downstream-blocker`: Blocking the downstream ecosystem to work properly 
- `p3-significant`: High-priority enhancement 

## General categories
- `feature`
- `bug`
- ~~`enhancement`~~
- `docs` : Improvements or additions to documentation
- `dx`
- `devtools`
- `test`
- `types`
- `performance`
- `cookbook`

## For Contributors
- `good first issue`: Good for newcomers
- `help wanted`: Extra attention is needed
- `contribution-welcome`
- `needs reproduction`

## Extra info
- `breaking-change`
- `regression`: The issue only appears after a new release
- `upstream`: Bug in a dependency of Tres
- `has-workaround`
- `cannot-reproduce`: The bug cannot be reproduced
- `duplicate`: This issue or pull request already exists
- `invalid`: This doesn't seem right
- `wontfix`: This will not be worked on
- `on-hold`: for whatever reason

## PR meta
- `needs-test`
- `needs-documentation`
- `needs-translations`

**Comment by alvarosabu** [comment]
@Tresjs/core please provide feedback to start this task

**Comment by kekkorider** [comment]
If not similar, we could apply the same exact label system.
Every label in that list can find its place in the TresJS ecosystem, so why not? 🤓

**Comment by andretchen0** [comment]
I think this would be useful.

I try to open issues when I come across them, but lots of them are non-breaking for most users and it'd be nice to signal that to the team in a systemic fashion. This system would allow me to do that.

## Proposal

For the purposes of clarity and alphabetical grouping of common themes, I'd prefer to see e.g.,

* bug-p2-edge-case
* bug-p3-minor
* bug-p4-important

then eliminate the "bug" label. 

Same for "feature".

* feat-p2-nice-to-have
* feat-p2-to-be-discussed

## Reasoning

The Github labels are shown alphabetically in the contexts where I run into them. I rather be presented with:

* bug-p2-edge-case
* bug-[...]
* bug-[...]
* feat-p2-nice-to-have
* feat-[...]
* feat-[...]

... instead of ...

* p2-edge-case <-- this is a bug
* p2-nice-to-have <-- this is a feature
* p3-[...] <-- bug or feature? not enough info

I think that will require less explanation for new contributors than the current Vite labels. And it'll make it easier for us to keep them separate.

## Context

This seems to be how the Vite team is using labels in practice. There are currently zero "is:open label:bug" or "is:open label:feature"

* https://github.com/vitejs/vite/labels/bug
* https://github.com/vitejs/vite/labels/feature

... but there's a lot of "is:open label:p2-nice-to-have"

https://github.com/vitejs/vite/labels/p2-nice-to-have

I think it'd be more useful to label those as 'bug-p2-nice-to-have', so that it's clear to everyone.

**Reply by alvarosabu** [reply]
Hey @andretchen0 thanks for the feedback.

Regarding adding the prefix, would that be necessary if the ticket already have the `feat` or `bug` label?

**Reply by patak-cat** [reply]
I think the prefix will make it more difficult to add the labels. We type `p` and we get all the priority labels first.

**Reply by andretchen0** [reply]
@patak-dev @alvarosabu 

Responding to @alvarosabu 's proposal as written above, as I understand it, I'd have to remember that:

* `p3-significant` is paired with `feat: *` not `bug: *`
* `p4-important` is paired `bug: *` and not `feature: *`

It's my weak preference not to have to remember that or explain it to others.

---

@patak-dev wrote:

> I think the prefix will make it more difficult to add the labels. We type p and we get all the priority labels first.

Sure thing. To bridge the gap, it sounds like if the priorities are either ...

* the same across `feat: *` and `bug: *` – e.g., `p3-significant` can be applied to a feature or a bug
* or clearly only applicable to `feat: *` or `bug: *` without further context – e.g., `p2-edge-case`
 
... then just using `p*` is perfect.

Typing `p` will bring up all the priority labels, and I don't have to remember which is applicable to the current issue.

---

@alvarosabu wrote:

> Regarding adding the prefix, would that be necessary if the ticket already have the feat or bug label?

It seems that the Vite folks are largely not using plain `bug` and `feature` labels. They are e.g., adding multiple `feat: *`. 

For example:

https://github.com/vitejs/vite/pull/16129

**Reply by alvarosabu** [reply]
@andretchen0 mmm it's true that `p3-significant` and `p4-important` are too vague to cognitively relate them to bug or feat while others like `p2-edge-case` wording already implies is related to a bug (there are no edge-cases features), similar to `p2-nice-to-have` (there are no nice to have bug fixes 😂)

> ... then just using p* is perfect.

> Typing p will bring up all the priority labels, and I don't have to remember which is applicable to the current issue.

Not quite, we would use this prio labels to categorize issues and PRs from the community, as a contributor, I would like to have some context of why my request is `p4` or `p1`, by just setting the number we are not giving clear feedback to the user why the featured was marked with that priority. 

It also helps when 2 different context has the same level of prio, for example `p2-nice-to-have` with `p2-needs-discussion`, both have same weight on prio but their nature is different.

**Reply by andretchen0** [reply]
Hey @alvarosabu 

Sure thing. I think the top-line proposal is an improvement. Go for it.

**Comment by andretchen0** [comment]
## Proposal

Eliminate "enhancement". Use "feat-*" instead.

## Reasoning

* The difference between "enhancement" and "feature" is subtle in many cases.
* Our [new issue page](https://github.com/Tresjs/tres/issues/new/choose) doesn't make the distinction. 
* We don't allow "enhancement:" commit messages per [Conventional commits](https://www.conventionalcommits.org/en/v1.0.0-beta.4/), but we do allow "feat:".

**Reply by alvarosabu** [reply]
Totally agree

**Comment by andretchen0** [comment]
## Proposal

Format labels using `kebab-case`. Without spaces, punctuation, etc.

## Reasoning

### Consistency

In the top-line proposal, there's a mix of spaces and hyphens. E.g.,:

* p2-edge-case
* bug: upstream

It'd be nice to have a single way of formatting labels.

### No escaping required

Depending on the context, spaces and punctuation have to be escaped.

"feat: runtime" can sometimes be escaped as ...

```
https://github.com/vitejs/vite/labels/feat%3A runtime
```

... but sometimes has to be escaped as ...

```
https://github.com/vitejs/vite/labels/feat%3A%20runtime
```

### No quoting required

A label with spaces sometimes has to be quoted ...

```
is:open label:"feat: runtime" 
```

### Meanwhile, kebab just works

Kebab-case labels work as written, afaik ...

```
https://github.com/vitejs/vite/labels/p3-significant
```

```
is:open label:p2-nice-to-have 
```

**Comment by alvarosabu** [comment]
> It'd be nice to have a single way of formatting labels.

👍.  Especially due the escaping, agree.

**Reply by alvarosabu** [reply]
The only issue is, some of them are using by Github to bring new contributors:

- `help wanted`
- `good first issue`

We will need to maintain those

**Comment by hawk86104** [comment]
Haha, I finally figured out how to submit a PR correctly.

**Comment by ThimoDEV** [comment]
Just my thoughts about the label/issue system. How do other projects let new or outside contributors of the core team know which issues they can pick up? Some issues or bugs which are edge cases for example can be perfect for a new contributor but if it salready worked on by a core team member then its wasted energy. Is it an idea to have a special label for that situation besides good first issue?

On a side note: How to make this clear in the discord if desired?

**Reply by alvarosabu** [reply]
Github has two default labels to promote tasks for new joiners, `good first issue` and `help wanted`, we can set a bot on the #contributing channel to provide the link for all the tasks under that category

**Reply by ThimoDEV** [reply]
That seems a great idea! The bot would have it pinned then or maybe a separate issues contribution channel where someone can assign themself to an issue in some way too
